### DISCUSSION TITLE
Confused by strict / nonstrict / parse behaviour & documentation

### DISCUSSION THREAD

**Discussion by frankiesardo** [colinhacks/zod#490] [Q&A]
I was wondering if it was possible to offer more extensive documentation about the strict, nostrict and strip behaviour provided by zod. The readme mentions the various methods, but it would be great to provide more exaples

**Comment by harshagarwalnyu** [comment]
Zod objects have three modes for handling **unknown keys** (keys not defined in the schema):

---

**`strip` (default)** — silently removes unknown keys from the output:
```ts
const Schema = z.object({ name: z.string() });

Schema.parse({ name: 'Alice', age: 30, extra: true });
// Output: { name: 'Alice' }  — age and extra are removed
```

---

**`strict`** — throws a `ZodError` if unknown keys are present:
```ts
const Schema = z.object({ name: z.string() }).strict();

Schema.parse({ name: 'Alice', age: 30 });
// Throws: ZodError — unrecognized key: "age"

Schema.parse({ name: 'Alice' });
// Output: { name: 'Alice' }  — OK
```
Use this when you want to guarantee no extra data leaks through (e.g., API input validation).

---

**`passthrough`** — keeps unknown keys in the output without validating them:
```ts
const Schema = z.object({ name: z.string() }).passthrough();

Schema.parse({ name: 'Alice', age: 30, extra: true });
// Output: { name: 'Alice', age: 30, extra: true }  — all keys preserved
```
The TypeScript type still only shows `{ name: string }` — unknown keys are runtime-only.

---

**`nonstrict`** — deprecated alias for `passthrough`. Don't use it in new code.

---

**Switching modes at runtime:**
```ts
const base = z.object({ name: z.string() });
base.strict()      // throws on unknown
base.strip()       // (default) strips unknown
base.passthrough() // keeps unknown
```

**TypeScript types:**
- `strip` / `strict` → `{ name: string }`
- `passthrough` → `{ name: string } & { [k: string]: unknown }` (not quite, but conceptually)

**When to use which:**
| Mode | Use case |
|------|----------|
| `strip` (default) | Most cases — safe default, prevents data leakage |
| `strict` | API gateway / deserialization where extra keys = bug |
| `passthrough` | Proxying data, extending schemas, when you need to forward unknown keys |
